### PR TITLE
Feature/task 287 celery request

### PR DESCRIPTION
**Pull request by annashamray**
Based on branch `feature/task276-request-exception`
PR - https://github.com/VNG-Realisatie/gemma-notificatiecomponent/pull/13 

**Changes:**
1. add celery app and config
1. add celery task `send_msg_to_sub_task`
1. replace loop of requests with tasks in `MessageSerializer`
1. incorporate tasks in tests

### CODE REVIEW

File: requirements/dev.in [hunk 1]
```diff
@@ -1 +1,2 @@
 django-debug-toolbar
+flower
```
Comment (sergei-maertens) [inline requirements/dev.in]: Probably good to also have this running on kubernetes/docker-compose

File: src/notifications/api/tests/test_tasks.py [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+        notif = NotificatieFactory.create()
+
+        request_data = {
+            "kanaal": "zaken",
+            "hoofdObject": "https://ref.tst.vng.cloud/zrc/api/v1/zaken/d7a22",
+            "resource": "status",
+            "resourceUrl": "https://ref.tst.vng.cloud/zrc/api/v1/statussen/d7a22/721c9",
+            "actie": "create",
+            "aanmaakdatum": "2018-01-01T17:00:00Z",
+            "kenmerken": [
```
Comment (sergei-maertens) [inline src/notifications/api/tests/test_tasks.py]: format here as well

File: src/notifications/celery.py [hunk 3]
```diff
@@ -0,0 +1,10 @@
+import os
+
+from celery import Celery
+
+os.environ.setdefault("DJANGO_SETTINGS_MODULE", "notifications.conf.dev")
```
Comment (sergei-maertens) [inline src/notifications/celery.py]: look at `src/notifications/wsgi.py` to setup the env - same mechanism should be used here

File: src/notifications/conf/base.py [hunk 4]
```diff
@@ -342,3 +342,7 @@
 # RabbitMQ
 BROKER_URL = os.getenv('PUBLISH_BROKER_URL', 'amqp://guest:guest@localhost:5672/%2F')
 CHANNEL = QueueChannel(params=BROKER_URL)
+
+# Celery
+CELERY_BROKER_URL = "amqp://127.0.0.1:5672//"
```
Comment (sergei-maertens) [inline src/notifications/conf/base.py]: these need to be pulled from the environment, with a fallback to defaults. See above ;)

File: src/notifications/conf/base.py [hunk 5]
```diff
@@ -342,3 +342,7 @@
...
+CELERY_RESULT_BACKEND = "amqp://127.0.0.1:5672//"
```
Comment (sergei-maertens) [inline src/notifications/conf/base.py]: This should be pulled from the env as well

File: src/notifications/api/serializers.py [hunk 6]
```diff
@@ -11,8 +11,9 @@
 from requests.exceptions import RequestException
```
[the remark this answers is not in the document]
Comment (annashamray) [reply src/notifications/api/serializers.py]: imports are removed

File: src/notifications/api/tasks.py [hunk 7]
```diff
@@ -0,0 +1,42 @@
+import requests
+
+from notifications.celery import app
+from notifications.datamodel.models import Abonnement, NotificatieResponse
+
+
+def send_msg_to_sub(sub_id, msg, notificatie_id):
```
[the remark this answers is not in the document]
Comment (annashamray) [reply src/notifications/api/tasks.py]: the function is renamed

[the remark this answers is not in the document]
Comment (annashamray) [reply src/notifications/api/tasks.py]: Thanks! Fixed

File: src/notifications/api/tasks.py [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+    """
```
[the remark this answers is not in the document]
Comment (annashamray) [reply src/notifications/api/tasks.py]: fixed

File: src/notifications/api/tasks.py [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+        send msg to subscriber
+        Write results to NotificatieResponse
+        """
+    sub = Abonnement.objects.get(id=sub_id)
```
[the remark this answers is not in the document]
Comment (annashamray) [reply src/notifications/api/tasks.py]: added try: except

File: src/notifications/api/tasks.py [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+        status_code = response.status_code
+    except requests.exceptions.RequestException as e:
+        # log of the response of the call
+        NotificatieResponse.objects.create(
+            notificatie_id=notificatie_id,
+            abonnement=sub,
+            exception=str(e)
+        )
+        status_code = None
+    return status_code
```
[the remark this answers is not in the document]
Comment (annashamray) [reply src/notifications/api/tasks.py]: removed return values from the task

File: src/notifications/api/tests/test_notificatie.py [hunk 11]
```diff
@@ -81,64 +81,25 @@ def test_notificatie_send_abonnement(self, mock_queue):
...
 
-        with requests_mock.mock() as m:
-            m.post(abon.callback_url)
-
-            response = self.client.post(notificatie_url, request_data)
+        response = self.client.post(notificatie_url, request_data)
 
         self.assertEqual(response.status_code, status.HTTP_201_CREATED, response.data)
         self.assertEqual(Notificatie.objects.count(), 1)
+        mock_task.assert_called_with(
```
[the remark this answers is not in the document]
Comment (annashamray) [reply src/notifications/api/tests/test_notificatie.py]: fixed

File: src/notifications/api/tests/test_notificatie.py [hunk 12]
```diff
@@ -165,50 +126,8 @@ def test_notificatie_send_abonnement_inconsistent_kenmerken(self, mock_queue):
...
 
-        with requests_mock.mock() as m:
-            m.post(abon.callback_url)
-
-            response = self.client.post(notificatie_url, request_data)
+        response = self.client.post(notificatie_url, request_data)
 
         self.assertEqual(response.status_code, status.HTTP_400_BAD_REQUEST)
         validation_error = response.data['kenmerken'][0]
         self.assertEqual(validation_error.code, 'kenmerken_inconsistent')
```
[the remark this answers is not in the document]
Comment (annashamray) [reply src/notifications/api/tests/test_notificatie.py]: added `mock_task.assert_not_called()`

File: src/notifications/api/tests/test_tasks.py [hunk 13]
```diff
@@ -0,0 +1,120 @@
...
+from notifications.datamodel.models import NotificatieResponse
+from notifications.datamodel.tests.factories import (
+    AbonnementFactory, NotificatieFactory
+)
+
+from ..tasks import send_msg_to_sub
+
+
+class NotifCeleryTests(APITestCase):
+    @patch('notifications.api.tasks.requests.post',
```
[the remark this answers is not in the document]
Comment (annashamray) [reply src/notifications/api/tests/test_tasks.py]: `mock.patch` is replaced with `request_mocks`

File: src/notifications/api/tests/test_tasks.py [hunk 14]
```diff
@@ -0,0 +1,120 @@
...
+from rest_framework import status
+from rest_framework.response import Response
+from rest_framework.test import APITestCase
+
+from notifications.datamodel.models import NotificatieResponse
+from notifications.datamodel.tests.factories import (
+    AbonnementFactory, NotificatieFactory
+)
+
+from ..tasks import send_msg_to_sub
```
[the remark this answers is not in the document]
Comment (annashamray) [reply src/notifications/api/tests/test_tasks.py]: Oh, thanks! I tried to write a test with a task and `.delay()` method but failed to integrate a worker with test DB. Therefore I decided to tetst a function directly

File: src/notifications/api/tests/test_tasks.py [hunk 15]
```diff
@@ -0,0 +1,120 @@
...
+        notif = NotificatieFactory.create()
+
+        request_data = {
+            "kanaal": "zaken",
+            "hoofdObject": "https://ref.tst.vng.cloud/zrc/api/v1/zaken/d7a22",
+            "resource": "status",
+            "resourceUrl": "https://ref.tst.vng.cloud/zrc/api/v1/statussen/d7a22/721c9",
+            "actie": "create",
+            "aanmaakdatum": "2018-01-01T17:00:00Z",
+            "kenmerken": [
```
[the remark this answers is not in the document]
Comment (annashamray) [reply src/notifications/api/tests/test_tasks.py]: fixed

File: src/notifications/api/tests/test_tasks.py [hunk 16]
```diff
@@ -0,0 +1,120 @@
...
+                {"bron": "082096752011"},
+                {"zaaktype": "example.com/api/v1/zaaktypen/5aa5c"},
+                {"vertrouwelijkheidaanduiding": "openbaar"}
+            ]
+        }
+
+        status_code = send_msg_to_sub(abon.id, request_data, notif.id)
```
[the remark this answers is not in the document]
Comment (annashamray) [reply src/notifications/api/tests/test_tasks.py]: fixed. Used `m.last_request` in assertions

File: src/notifications/api/tests/test_tasks.py [hunk 17]
```diff
@@ -0,0 +1,120 @@
...
+        self.assertEqual(status_code, status.HTTP_201_CREATED)
+
+        mock_args, mock_kwargs = mock_send.call_args
+
+        self.assertEqual(mock_args[0], abon.callback_url)
+        self.assertEqual(mock_kwargs['data'], request_data)
+        self.assertEqual(mock_kwargs['headers']['Content-Type'], 'application/json')
+        self.assertEqual(mock_kwargs['headers']['Authorization'], abon.auth)
+
+    @patch('notifications.api.tasks.requests.post',
```
[the remark this answers is not in the document]
Comment (annashamray) [reply src/notifications/api/tests/test_tasks.py]: fixed

File: src/notifications/api/tests/test_tasks.py (hunk 2 above)
Comment (annashamray) [reply src/notifications/api/tests/test_tasks.py]: fixed

File: src/notifications/celery.py (hunk 3 above)
Comment (annashamray) [reply src/notifications/celery.py]: fixed

File: src/notifications/conf/base.py (hunk 4 above)
Comment (annashamray) [reply src/notifications/conf/base.py]: fixed. Used `os.getenv`

File: src/notifications/conf/base.py (hunk 5 above)
Comment (annashamray) [reply src/notifications/conf/base.py]: fixed. Used os.getenv

**Pull request merged**